### PR TITLE
Fix `string::replace` under ASan

### LINKED ISSUE #3883: `<string>`: `basic_string::replace` fails under ASan when `pos + count > size`, and `count2 < count`
```cxx
#include <string>

int main()
{
  std::string t = "0123456789ABCDEF"; // large string
  t.replace(0, 30, 7, 'A'); // Address Sanitizer Error: Container overflow.
}
```

Looks like we (I) screwed up the logic here.

Reported in DevCom-10369462

### PR DESCRIPTION
**Pull request by strega-nil-ms**
Fixes #3883

### CODE REVIEW

File: tests/std/tests/GH_003883_replace_asan_failure/env.lst [hunk 1]
```diff
@@ -0,0 +1,57 @@
+# Copyright (c) Microsoft Corporation.
```
Comment (CaseyCarter) [inline tests/std/tests/GH_003883_replace_asan_failure/env.lst:1]: Could we add the test case to the existing string ASan test so we don't need to duplicate this giant `env.lst`? If not, then can we avoid duplication by promoting this up a directory, naming it `usual_asan_something`, and having the three asan tests share it?

Comment (StephanTLavavej) [reply tests/std/tests/GH_003883_replace_asan_failure/env.lst:1]: The string and vector `env.lst`s differ somewhat. I think we should fuse this into the existing string test that already tests a bugfix. I can validate and push changes.

File: tests/std/tests/GH_003883_replace_asan_failure/test.cpp [hunk 2]
```diff
@@ -0,0 +1,9 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <string>
+
+int main() {
+    std::string t = "0123456789ABCDEF"; // large string
+    t.replace(0, 30, 7, 'A');
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_003883_replace_asan_failure/test.cpp:8]: In general, after calling a modifying function, we should verify the final state. This helps catch all sorts of bugs, including those we might not have expected.

File: tests/std/tests/GH_003883_replace_asan_failure/env.lst (hunk 1 above)
Comment (StephanTLavavej) [reply tests/std/tests/GH_003883_replace_asan_failure/env.lst:1]: Also, there was a missing `test.lst` change here.

File: stl/inc/xstring [hunk 3]
```diff
@@ -3807,12 +3807,13 @@ public:
             // either we are shrinking, or the growth fits
             // may temporarily overflow; OK because size_type must be unsigned
             const auto _New_size = _Old_size + _Count - _Nx;
-            _ASAN_STRING_MODIFY(*this, _Old_size, _New_size);
+            _ASAN_STRING_REMOVE(*this);
             _Mypair._Myval2._Mysize = _New_size;
```
Comment (StephanTLavavej) [inline stl/inc/xstring:3811]: No change requested: We could fuse `_New_size` now, or move its definition below the `_ASAN_STRING_REMOVE`. This is fine as-is though.

**Review by StephanTLavavej — APPROVED** [review]

**Review by zacklj89 — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for this important ASAN fix! :lady_beetle: :hammer_and_wrench: :tada:

**Pull request merged**